Sparkle/SUBinaryDeltaApply.m

```diff
@@ -74,21 +130,42 @@ int applyBinaryDelta(NSString *source, NSString *destination, NSString *patchFil
         NSString *destinationFilePath = [destination stringByAppendingPathComponent:path];
 
         const char *value;
-        if (!xar_prop_get(file, "delete", &value) || !xar_prop_get(file, "delete-then-extract", &value)) {
-            removeTree(destinationFilePath);
-            if (!xar_prop_get(file, "delete", &value))
+        if (!xar_prop_get(file, DELETE_KEY, &value) || !xar_prop_get(file, DELETE_THEN_EXTRACT_KEY, &value)) {
+            if (!removeTree(destinationFilePath)) {
+                fprintf(stderr, "%s or %s: failed to remove %s\n", DELETE_KEY, DELETE_THEN_EXTRACT_KEY, [destination fileSystemRepresentation]);
+                return 1;
+            }
+            if (!xar_prop_get(file, DELETE_KEY, &value))
                 continue;
         }
 
-        if (!xar_prop_get(file, "binary-delta", &value))
-            applyBinaryDeltaToFile(x, file, sourceFilePath, destinationFilePath);
-        else
-            xar_extract_tofile(x, file, [destinationFilePath fileSystemRepresentation]);
+        if (!xar_prop_get(file, BINARY_DELTA_KEY, &value)) {
+            if (!applyBinaryDeltaToFile(x, file, sourceFilePath, destinationFilePath)) {
+                fprintf(stderr, "Unable to patch %s to destination %s\n", [sourceFilePath fileSystemRepresentation], [destinationFilePath fileSystemRepresentation]);
+                return 1;
+            }
+        } else if (xar_prop_get(file, MODIFY_PERMISSIONS_KEY, &value)) {
```

Comment:
File permissions changes can only happen:
a) After a binary diff is done, if permissions have changed
b) if the file's content does not change, but the permissions do change

They don't happen when a new file is extracted, because the new file's permissions from the xar will just be used. The confusion might stem from having to deal with delete and delete-then-extract properties.